### PR TITLE
Freezer: reimplement "now" patching with forbiddenfruit.

### LINKED ISSUE #22: Freezing: use forbiddenfruit for patching "now"?
The current implementation of `freeze` is quite heavy since we are replacing the `datetime` module and class in order to replace the `now` method.
This has side-effects such as wrong `isinstance` check results or pickling errors.

We could use https://github.com/clarete/forbiddenfruit for replacing `now`, so that we only replace the method and keep the class.
This would make a lot quite simpler.

### PR DESCRIPTION
**Pull request by jone**
This fixes problems with pickling and comparison of frozen datetime objects.

Closes #22 

/cc @mbaechtold

### CODE REVIEW

File: ftw/testing/freezer.py [hunk 1]
```diff
@@ -67,51 +52,12 @@ def now(tz=None):
...
-        if not self.enabled:
-            return
-
-        self.__exit__(None, None, None)
-        transaction.get().addAfterCommitHook(
-            self.transaction_after_commit_hook)
-
-    def transaction_after_commit_hook(self, status):
-        self.__enter__()
+        curse(datetime, 'now', self._previous_datetime_now)
```
Comment (lukasgraf) [inline ftw/testing/freezer.py]: Wouldn't [`forbiddenfruit.reverse()`](https://github.com/clarete/forbiddenfruit#reversing-a-curse) be more appropriate here (assuming it works)?

Comment (jone) [inline ftw/testing/freezer.py]: `reverse` is more like a delete: `now` is not there anymore afterwards, so it does not work.

Comment (lukasgraf) [inline ftw/testing/freezer.py]: Ah, I see. :+1:

**Comment by lukasgraf** [comment]
Looks good to me, aside from those weird test failures. What the heck is up with this?

```pytb
   File "/var/lib/jenkins/zope/eggs/Zope2-2.13.15-py2.6.egg/Shared/DC/xml/ppml.py", line 250, in ToXMLUnpickler
     dispatch[PERSID] = load_persid
 NameError: name 'PERSID' is not defined
```

`PERSID` should be imported from the `pickle` module.

**Comment by jone** [comment]
It works locally. I have no idea what is happening. The `master` still works.

**Comment by lukasgraf** [comment]
Hmm, your branch is also failing for me locally in the same way.

**Comment by lukasgraf** [comment]
Well, it definitely is related to `forbiddenfruit`. As soon as I remove the `from forbiddenfruit import curse` import, the layer setup stops failing. It somehow seems to affect the `from pickle import *` star import. If I change that to `from pickle import PERSID` in `ppml.py`, `PERSID` is defined (but the next name from the star import fails).

**Comment by lukasgraf** [comment]
I'm pretty confident this is an interaction between [`forbiddenfruit`'s `__filtered_dir__`](https://github.com/clarete/forbiddenfruit/blob/master/forbiddenfruit/__init__.py#L56-L64) and the fact that the Python `pickle` module first defines a [limited `__all__`](https://hg.python.org/cpython/file/2.7/Lib/pickle.py#l37), but then [later extends it using `dir()`](https://hg.python.org/cpython/file/2.7/Lib/pickle.py#l165).

**Comment by lukasgraf** [comment]
@jone Yeah, the problem is that [`forbiddenfruit`'s `__filtered_dir__`](https://github.com/clarete/forbiddenfruit/blob/master/forbiddenfruit/__init__.py#L56-L64) breaks the use of `dir()` without arguments to inspect the locals,  [as used by the `pickle` module`](https://hg.python.org/cpython/file/2.7/Lib/pickle.py#l165).

Because of the way `dir()` is implemented in CPython (see [`_dir_locals`](https://hg.python.org/cpython/file/2.7/Objects/object.c#l1788) and [`PyObject_Dir`](https://hg.python.org/cpython/file/2.7/Objects/object.c#l1968)), `_dir_locals()` will always return the local scope of the calling frame. In the case of `forbiddenfruit`, this will always be the local scope of `__filtered_dir__` because of the added indirection.

I'll see if I can create a decent bug report, but for now I don't think we can use FF - that side effect could cause some major headaches, even if it's just in tests.

**Comment by lukasgraf** [comment]
@jone addressed in clarete/forbiddenfruit#13

**Comment by jone** [comment]
@lukasgraf @mbaechtold The time has come! :sparkles:

**Comment by lukasgraf** [comment]
Let's give this a try!

File: docs/HISTORY.txt [hunk 2]
```diff
@@ -5,10 +5,8 @@ Changelog
 1.10.3 (unreleased)
 -------------------
 
-- Freezer: disable freezing while committing to database for preventing pickling errors.
-  [jone]
-
-- Freezer bugfix: replace datetime instances when leaving freeze context manager.
+- Freezer: reimplement "now" patching with forbiddenfruit.
+  This fixes problems with pickling and comparison of frozen datetime objects.
```
Comment (lukasgraf) [inline docs/HISTORY.txt]: This changelog got messed up a bit I think - `1.10.3` has been released in the meantime (containing "disable while committing to db" and "... when leaving context manager"), but this entry replaced the other two retroactively in the existing `1.10.3` release notes.

@jone Could you take a look at that and fix it? (My suggestion would be to simply re-add the old entries to 1.10.3, and move the new entry to a 1.11.0 section)

Comment (jone) [inline docs/HISTORY.txt]: :+1: thanks, fixed in master

**Pull request merged**